File: src/net/eduard/api/lib/menu/Shop.java
package net.eduard.api.lib.menu;

import java.util.*;
import java.util.function.Consumer;
import java.util.stream.Stream;

import net.eduard.api.lib.game.FakePlayer;
import net.eduard.api.lib.modules.*;
import net.eduard.api.server.currency.CurrencyController;
import net.eduard.api.server.currency.CurrencyHandler;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.inventory.ClickType;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;

import net.eduard.api.lib.manager.CurrencyManager;
import net.eduard.api.lib.storage.Storable.*;

@StorageAttributes(indentificate = true)
public class Shop extends Menu {


    public static String MESSAGE_CHOICE_AMOUNT = "§aEscolha uma quantidade para $trader este produto $product";
    public static String MESSAGE_BOUGHT_ITEM = "§aVoce adquiriu $amount ($product) produto(s) da Loja!";
    public static String MESSAGE_SOLD_ITEM = "§aVocê vendeu $amount ($product) produtos(s) para a loja!";
    public static String MESSAGE_WITHOUT_ITEMS = "§cVoce não tem items suficiente!";
    public static String MESSAGE_WITHOUT_BALANCE = "§cVoce não tem dinheiro suficiente!";
    public static String MESSAGE_WITHOUT_PERMISSION = "§cVoce não tem permissão para comprar este produto!";
    public static List<String> TEMPLATE_BUY = Arrays.asList("§fCompre o produto §e$product_name",
            "§2Quantidade: §a$product_stock", "§2Preço por 1: §a$product_buy_unit_price", "§2Preço por 64: §a$product_buy_pack_price");

    public static List<String> TEMPLATE_SELL = Arrays.asList("§fVende o produto: §e$product_name",
            "§2Quantidade: §a$product_stock", "§2Preço por 64: §a$product_sell_pack_price", "§2Preço por Inventario: §a$product_sell_inventory_price");
    public static List<String> TEMPLATE_BUY_SELL = Arrays.asList("§fCompra e venda de: §e$product_name",
            "§2Quantidade: §a$product_stock", "§2Preço por 64: §a$product_sell_pack_price", "§2Preço por Inventario: §a$product_sell_inventory_price", "", "§2Preço por 1: §a$product_buy_unit_price", "§2Preço por 64: §a$product_buy_pack_price");


    private String currencyType = "VaultEconomy";

    transient private CurrencyHandler currency;
    private static final int PLAYER_INVENTORY_LIMIT = 4 * 64 * 9;
    private ShopSortType sortType = ShopSortType.BUY_PRICE_ASC;
    private transient Map<Player, Product> selectingAmount = new HashMap<>();
    private transient Map<Player, TradeType> trading = new HashMap<>();
    private boolean amountPerChat;
    private List<String> buyTemplate = new ArrayList<>(TEMPLATE_BUY);
    private List<String> sellTemplate = new ArrayList<>(TEMPLATE_SELL);
    private List<String> sellBuyTemplate = new ArrayList<>(TEMPLATE_BUY_SELL);

    private String messageChoiceAmount = MESSAGE_CHOICE_AMOUNT;
    private String messageBoughtItem = MESSAGE_BOUGHT_ITEM;
    private String messageSoldItem = MESSAGE_SOLD_ITEM;
    private String messageWithoutItems = MESSAGE_WITHOUT_ITEMS;
    private String messageWithoutBalance = MESSAGE_WITHOUT_BALANCE;
    private String messageWithoutPermission = MESSAGE_WITHOUT_PERMISSION;


    public Shop() {
        this("Menu", 3);
    }

    public Shop copy() {
        return copy(this);
    }

    public void organize() {

        if (sortType == ShopSortType.BUY_PRICE_ASC) {
            Stream<Product> lista = getButtons().stream().filter(b -> b instanceof Product).map(b -> (Product) b)
                    .sorted(Comparator.comparing(Product::getUnitBuyPrice));
            lista.forEach(new Consumer<Product>() {
                int id = 0;

                public void accept(Product t) {
                    t.setIndex(id);
                    id++;
                }
            });
        }

    }

    @EventHandler
    public void chat(AsyncPlayerChatEvent e) {
        Player p = e.getPlayer();
        if (selectingAmount.containsKey(p)) {
            Product product = selectingAmount.get(p);
            Double amount = Extra.fromMoneyToDouble(e.getMessage());
            amount = Math.abs(amount);
            selectingAmount.remove(p);
            TradeType trade = trading.get(p);
            trading.remove(p);
            if (trade == TradeType.BUYABLE) {
                buy(p, product, amount);
            } else if (trade == TradeType.SELABLE) {
                sell(p, product, amount);
            }
            e.setCancelled(true);

        }
    }

    public Shop(String title, int lineAmount) {
        super(title, lineAmount);
        setEffect((event, page) -> {
            if (event.getWhoClicked() instanceof Player) {
                Player player = (Player) event.getWhoClicked();
                if (event.getCurrentItem() == null)
                    return;

                Product product = getProduct(event.getCurrentItem());
                if (product == null) {
                    return;
                }

                if ((event.getClick() == ClickType.RIGHT || event.getClick() == ClickType.SHIFT_RIGHT)
                        && (product.getTradeType() == TradeType.BOTH
                        || product.getTradeType() == TradeType.BUYABLE)) {


                    if (isAmountPerChat()) {
                        selectingAmount.put(player, product);
                        trading.put(player, TradeType.BUYABLE);
                        player.closeInventory();
                        player.sendMessage(messageChoiceAmount.replace("$product",
                                "" + product.getName()).replace("$trader", "comprar"));
                        return;
                    }

                    int amount = 1;
                    if (event.getClick() == ClickType.SHIFT_RIGHT) {
                        amount = product.getAmount();
                        if (amount < 64) {
                            amount = 64;
                        }

                    }
                    buy(player, product, amount);


                }

                if ((event.getClick() == ClickType.LEFT || event.getClick() == ClickType.SHIFT_LEFT)
                        && (product.getTradeType() == TradeType.BOTH
                        || product.getTradeType() == TradeType.SELABLE)) {
                    if (product.getProduct() != null) {
                        if (isAmountPerChat()) {
                            selectingAmount.put(player, product);
                            trading.put(player, TradeType.SELABLE);
                            player.closeInventory();
                            player.sendMessage(messageChoiceAmount.replace("$product",
                                    "" + product.getName()).replace("$trader", "vender"));
                            return;
                        }

                        int amount = Mine.getTotalAmount(player.getInventory(), product.getProduct());
                        if (event.getClick() == ClickType.LEFT) {
                            if (amount > 64) {
                                amount = 64;
                            }
                        }
                        sell(player, product, amount);

                    }

                }

            }

        });

        // TODO Auto-generated constructor stub
    }


    public void buy(Player player, Product product, double amount) {
        FakePlayer fake = new FakePlayer(player);

        double priceUnit = product.getUnitBuyPrice();
        if (product.isLimited() && amount > product.getStock()) {
            amount = product.getStock();
        }
        double priceFinal = priceUnit * amount;
        ProductTradeEvent evento = new ProductTradeEvent(player);
        evento.setProduct(product);
        evento.setAmount(amount);
        if (product.isLimited()) {
            evento.setNewStock(product.getStock() - amount);
        } else {
            evento.setNewStock(product.getStock());
        }
        ;


        evento.setBalance(getCurrency().get(fake));


        evento.setType(TradeType.BUYABLE);
        evento.setPriceTotal(priceFinal);
        evento.setShop(Shop.this);

        Mine.callEvent(evento);
        if (evento.isCancelled()) {
            return;
        }

        if (getCurrency().get(fake) >= evento.getPriceTotal()) {
            getCurrency().remove(fake, evento.getPriceTotal());
        } else {
            player.sendMessage(messageWithoutBalance);
            return;
        }

        product.setStock(evento.getNewStock());

        for (String cmd : product.getCommands()) {
            Mine.runCommand(cmd.replace("$player", player.getName()).replace("$formated_amount", Extra.formatMoney(amount)).replace("$amount", "" + amount));
        }
        player.sendMessage(messageBoughtItem.replace("$amount", Extra.formatMoney(amount)).replace("$product",
                "" + product.getName()));
        if (product.getProduct() == null) {
            return;
        }

        ItemStack clone = product.getProduct().clone();
        if (evento.getAmount() > PLAYER_INVENTORY_LIMIT) {
            clone = MineReflect.toStack(clone, evento.getAmount());


        } else {
            clone.setAmount((int) evento.getAmount());
        }

        if (Mine.isFull(player.getInventory())) {
            Inventory inv = Mine.newInventory("Pegue seus items comprados", 6 * 9);
            inv.addItem(clone);
            player.openInventory(inv);
            player.sendMessage("§cPegue seus items comprados e coloca no seu inventário.");

            //player.getWorld().dropItemNaturally(player.getLocation().add(0, 5, 0),
            //      clone);
        } else {
            player.getInventory().addItem(clone);
        }
    }

    public void sell(Player player, Product product, double amount) {
        FakePlayer fake = new FakePlayer(player);
        double priceUnit = product.getUnitSellPrice();


        if (amount > product.getStock()) {
            amount = product.getStock();
        }
        if (amount < 1) {
            player.sendMessage(messageWithoutItems);
            return;
        }
        double finalPrice = amount * priceUnit;
        ProductTradeEvent evento = new ProductTradeEvent(player);
        evento.setProduct(product);
        evento.setAmount(amount);
        evento.setNewStock(product.getStock() - amount);

        evento.setBalance(getCurrency().get(fake));


        evento.setType(TradeType.SELABLE);
        evento.setPriceTotal(finalPrice);
        evento.setShop(Shop.this);
//
        Mine.callEvent(evento);
        if (evento.isCancelled()) {
            return;
        }
        if (evento.getAmount() > PLAYER_INVENTORY_LIMIT) {
            evento.setAmount(PLAYER_INVENTORY_LIMIT);
        }
        Mine.remove(player.getInventory(), product.getProduct(), (int) evento.getAmount());
        player.sendMessage(messageSoldItem.replace("$amount", "" + amount).replace("$product",
                "" + product.getName()));

        getCurrency().add(fake, finalPrice);


    }

    public CurrencyHandler getCurrency() {

        if (currency == null) {
            currency = CurrencyController.getInstance().getCurrencyHandler(currencyType);
        }

        return currency;
    }

    public String getMessageBoughtItem() {
        return messageBoughtItem;
    }

    public String getMessageWithoutBalance() {
        return messageWithoutBalance;
    }

    public Product getProduct(ItemStack icon) {
        MenuButton button = getButton(icon);
        if (button != null) {
            if (button instanceof Product) {
                Product product = (Product) button;
                return product;
            }
        }
        return null;
    }

    public Product getProductFrom(ItemStack item) {
        for (MenuButton button : getButtons()) {
            if (button instanceof Product) {
                Product product = (Product) button;
                if (product.getItem().isSimilar(item))
                    return product;
            }

        }
        return null;
    }

    public void setCurrency(CurrencyManager currency) {
        if (currency == null)
            return;

        this.currency = currency;
        for (MenuButton button : getButtons()) {
            if (button.getMenu() instanceof Shop) {
                Shop shop = (Shop) button.getMenu();
                shop.setCurrency(currency);

            }

        }
    }

    public void setMessageBoughtItem(String messageBoughtItem) {
        this.messageBoughtItem = messageBoughtItem;
    }

    public void setMessageWithoutBalance(String messageWithoutBalance) {
        this.messageWithoutBalance = messageWithoutBalance;
    }


    public String getMessageWithoutPermission() {
        return messageWithoutPermission;
    }

    public void setMessageWithoutPermission(String messageWithoutPermission) {
        this.messageWithoutPermission = messageWithoutPermission;
    }

    public ShopSortType getSortType() {
        return sortType;
    }

    public void setSortType(ShopSortType sortType) {
        this.sortType = sortType;
    }

    public String getMessageSoldItem() {
        return messageSoldItem;
    }

    public void setMessageSoldItem(String messageSoldItem) {
        this.messageSoldItem = messageSoldItem;
    }

    public String getMessageWithoutItems() {
        return messageWithoutItems;
    }

    public void setMessageWithoutItems(String messageWithoutItems) {
        this.messageWithoutItems = messageWithoutItems;
    }

    public boolean isAmountPerChat() {
        return amountPerChat;
    }

    public void setAmountPerChat(boolean amountPerChat) {
        this.amountPerChat = amountPerChat;
    }

    public String getMessageChoiceAmount() {
        return messageChoiceAmount;
    }

    public void setMessageChoiceAmount(String messageChoiceAmount) {
        this.messageChoiceAmount = messageChoiceAmount;
    }

    public Map<Player, TradeType> getTrading() {
        return trading;
    }

    public void setTrading(Map<Player, TradeType> trading) {
        this.trading = trading;
    }

    public List<String> getBuyTemplate() {
        return buyTemplate;
    }

    public void setBuyTemplate(List<String> buyTemplate) {
        this.buyTemplate = buyTemplate;
    }

    public List<String> getSellTemplate() {
        return sellTemplate;
    }

    public void setSellTemplate(List<String> sellTemplate) {
        this.sellTemplate = sellTemplate;
    }

    public List<String> getSellBuyTemplate() {
        return sellBuyTemplate;
    }

    public void setSellBuyTemplate(List<String> sellBuyTemplate) {
        this.sellBuyTemplate = sellBuyTemplate;
    }

    public String getCurrencyType() {
        return currencyType;
    }

    public void setCurrencyType(String currencyType) {
        this.currencyType = currencyType;
    }
}

File: src/main/java/net/eduard/api/lib/menu/Shop.kt
package net.eduard.api.lib.menu

import net.eduard.api.lib.game.ItemBuilder
import net.eduard.api.lib.kotlin.player
import net.eduard.api.server.currency.CurrencyManager
import net.eduard.api.lib.modules.*
import net.eduard.api.lib.plugin.IPluginInstance
import net.eduard.api.lib.storage.annotations.StorageAttributes

import net.eduard.api.server.CurrencySystem
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.inventory.ClickType
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.inventory.ItemStack
import kotlin.math.abs

@Suppress("unused")
@StorageAttributes(indentificate = true)
open class Shop(
    name: String,
    lineAmount: Int
) : Menu(name, lineAmount) {

    constructor() : this("Loja", 3)

    inline fun product(name: String = "Produto", setup: (Product.() -> Unit)): Product {
        val product = Product(name, this)
        setup.invoke(product)
        return product
    }


    var currencyType = "VaultEconomy"

    @Transient
    var currency: CurrencySystem? = null
        set(value) {
            field = value
            for (button in buttons) {
                if (button.isCategory && button.menu is Shop) {
                    val shop = button.shop
                    shop.currency = value
                }
            }
        }
        get() {
            if (field == null) {
                field = CurrencyManager.getCurrency(currencyType)
            }
            return field
        }
    var sortable = false
    var sortType = ShopSortType.BUY_PRICE_ASC
    var menuUpgrades: Menu? = null
    var menuConfirmation: Menu? = null

    @Transient
    var selectingAmount = mutableMapOf<Player, Product>()

    @Transient
    var trading = mutableMapOf<Player, TradeType>()

    @Transient
    val selectedProduct = mutableMapOf<Player, Product>()

    var isAmountPerChat = false
    var isPermissionShop = false
    var glowIconWhenAlreadyBought = true
    var buyTemplate = TEMPLATE_BUY.toMutableList()
    val boughtTemplate = TEMPLATE_BOUGHT.toMutableList()
    var sellTemplate = TEMPLATE_SELL.toMutableList()
    var sellBuyTemplate = TEMPLATE_BUY_SELL.toMutableList()

    var messageChoiceAmount = MESSAGE_CHOICE_AMOUNT
    var messageBoughtItem = MESSAGE_BOUGHT_ITEM
    var messageSoldItem = MESSAGE_SOLD_ITEM
    var messageAlreadyBought = MESSAGE_ALREADY_BOUGHT
    var messageWithoutItems = MESSAGE_WITHOUT_ITEMS
    var messageWithoutBalance = MESSAGE_WITHOUT_BALANCE
    var messageWithoutPermission = MESSAGE_WITHOUT_PERMISSION
    var messageUpgradeBought = MESSAGE_UPGRADE_BOUGHT
    var messageUpgradesAlreadyBought = MESSAGE_UPGRADES_ALREADY_BOUGHT


    companion object {
        var MESSAGE_CHOICE_AMOUNT = "§aEscolha uma quantidade para %trader este produto %product"
        var MESSAGE_BOUGHT_ITEM = "§aVoce adquiriu %amount (%product) produto(s) da Loja!"
        var MESSAGE_SOLD_ITEM = "§aVocê vendeu %amount (%product) produtos(s) para a loja!"
        var MESSAGE_WITHOUT_ITEMS = "§cVoce não tem items suficiente!"
        var MESSAGE_WITHOUT_BALANCE = "§cVoce não tem dinheiro suficiente!"
        var MESSAGE_WITHOUT_PERMISSION = "§cVoce não tem permissão para comprar este produto!"
        var MESSAGE_ALREADY_BOUGHT = "§cVocê já comprou este produto."
        var MESSAGE_UPGRADE_BOUGHT = "§cVocê comprou evolução do %product para o nível %level"
        var MESSAGE_UPGRADES_ALREADY_BOUGHT = "§cTodos os upgrades já foram adquiridos."


        var TEMPLATE_BUY = listOf(
            "§e%product_name",
            "",
            "§7Custo de x1: §a%product_buy_unit_price",
            "§7Custo de x%product_amount: §a%product_buy_x_price",
            "§7Custo de x64: §a%product_buy_pack_price",
            "§7Estoque de x§b%product_stock"
        )
        var TEMPLATE_BOUGHT = listOf(
            "",
            "§a§lDESBLOQUEADO"
        )
        var TEMPLATE_SELL = listOf(
            "§e%product_name",
            "",
            "§7Valor de x64: §a%product_sell_pack_price",
            "§7Valor de x2304: §a%product_sell_inventory_price",
            "§7Estoque de: x§b%product_stock"
        )
        var TEMPLATE_BUY_SELL = listOf(
            "§e%product_name",
            "",
            "§cComprar",
            " §7Custo de x1: §a%product_buy_unit_price",
            " §7Custo de x%product_amount: §a%product_buy_x_price",
            " §7Custo de x64: §a%product_buy_pack_price",
            "",
            "§cVender",
            " §7Valor de x64: §a%product_sell_pack_price",
            " §7Valor de x2304: §a%product_sell_inventory_price",
            "",
            "§7Estoque de x§b%product_stock"
            )
        const val PLAYER_INVENTORY_LIMIT = 4 * 64 * 9
    }


    fun useUpgradesMenu() {
        menuUpgrades = menu("Lista de Upgrades", 6) {
            cantBeOpened()

            button("upgrade") {
                setPosition(3, 2)
                icon = ItemBuilder(Material.ANVIL).name("§aEvoluir")
            }


            button("product") {
                setPosition(5, 2)
                icon = ItemBuilder(Material.STONE)
            }

        }
    }

    fun useConfirmationMenu() {
        this.menuConfirmation = menu("Confirmar Transação", 3) {
            cantBeOpened()
            button("confirmar") {
                setPosition(3, 2)
                icon = ItemBuilder(Material.WOOL)
                    .data(5)
                    .name("§a§lCONFIRMAR")
                    .lore("§aClique para confirmar a transação.")
            }

            button("cancelar") {
                icon = ItemBuilder(Material.WOOL)
                    .data(14)
                    .name("§c§lCANCELAR")
                    .lore("§cClique para cancelar a transação.")
                setPosition(7, 2)
            }

            button("product") {
                setPosition(5, 2)
                icon = ItemBuilder(Material.STONE)
            }
        }

    }


    override fun copy(): Shop {
        return Copyable.copyObject(this)
    }

    fun organize() {
        if (!sortable)return
        if (sortType === ShopSortType.BUY_PRICE_ASC) {
            updateButtonPositions(buttons.filterIsInstance<Product>().sortedBy { it.buyPrice })
        } else if (sortType === ShopSortType.BUY_PRICE_DESC) {
            updateButtonPositions(buttons.filterIsInstance<Product>().sortedByDescending { it.buyPrice })
        } else if (sortType === ShopSortType.SELL_PRICE_ASC) {
            updateButtonPositions(buttons.filterIsInstance<Product>().sortedBy { it.sellPrice })
        } else if (sortType === ShopSortType.SELL_PRICE_DESC) {
            updateButtonPositions(buttons.filterIsInstance<Product>().sortedByDescending { it.sellPrice })
        }
    }

    @EventHandler
    fun chat(e: AsyncPlayerChatEvent) {
        val player = e.player
        if (selectingAmount.containsKey(player)) {
            val product = selectingAmount[player]!!
            e.isCancelled = true
            var amount = Extra.fromMoneyToDouble(e.message)
            amount = abs(amount)
            selectingAmount.remove(player)
            val trade = trading[player]
            trading.remove(player)
            if (trade === TradeType.BUYABLE) {
                buy(player, product, amount)
            } else if (trade === TradeType.SELABLE) {
                sell(player, product, amount)
            }

        }
    }

    fun buy(player: Player, product: Product, value: Double) {
        var amount = value
        val fake = FakePlayer(player)
        val priceUnit = product.unitBuyPrice
        if (product.isLimited && amount > product.stock) {
            amount = product.stock
        }
        if (isPermissionShop) {
            if (player.hasPermission(product.permission)) {
                player.sendMessage(messageAlreadyBought)
                return
            }
        } else {
            if (!player.hasPermission(product.permission)) {
                player.sendMessage(messageWithoutPermission)
                return
            }
        }
        val priceFinal = priceUnit * amount
        val evento = ProductTradeEvent(player)
        evento.product = product
        evento.amount = amount
        if (product.isLimited) {
            evento.newStock = product.stock - amount
        } else {
            evento.newStock = product.stock
        }
        evento.balance = currency!!.get(fake)
        evento.type = TradeType.BUYABLE
        evento.priceTotal = priceFinal
        evento.shop = this@Shop
        Mine.callEvent(evento)
        if (evento.isCancelled) {
            debug("Cancelando a troca")
            return
        }
        if (currency!!.contains(fake, evento.priceTotal)) {
            currency!!.remove(fake, evento.priceTotal)
        } else {
            player.sendMessage(messageWithoutBalance)
            return
        }
        // após uma compra ser feita define o stock do evento para o produto
        product.stock = evento.newStock
        for (cmd in product.commands) {
            val cmdExecuted = cmd.replace("%player", player.name)
                .replace("%formated_amount", Extra.formatMoney(amount))
                .replace("%amount", "" + amount)
            debug("CMD: $cmdExecuted")
            Mine.runCommand(cmdExecuted)
        }

        player.sendMessage(
            messageBoughtItem
                .replace("%amount", Extra.formatMoney(amount))
                .replace("%product", product.display)
        )

        if (isPermissionShop) {
            if (VaultAPI.hasVault() && VaultAPI.hasPermission()) {
                VaultAPI.getPermission().playerAdd(null, fake, product.permission)
            }
            return
        }
        if (product.product == null) {
            return
        }
        var productItem = product.product!!.clone()
        if (evento.amount > PLAYER_INVENTORY_LIMIT) {
            productItem = MineReflect.toStack(productItem, evento.amount)
        } else {
            productItem.amount = evento.amount.toInt()
        }
        if (Mine.isFull(player.inventory)) {
            val inventory = Mine.newInventory("Pegue seus items comprados", 6 * 9)
            inventory.addItem(productItem)
            player.openInventory(inventory)
            player.sendMessage("§cPegue seus items comprados e coloca no seu inventário.")
        } else {
            player.inventory.addItem(productItem)
        }
    }

    fun sell(player: Player, product: Product?, value: Double) {
        var amount = value
        val fake = FakePlayer(player)
        val priceUnit = product!!.unitSellPrice
        if (amount > product.stock) {
            amount = product.stock
        }
        if (amount < 1) {
            player.sendMessage(messageWithoutItems)
            return
        }
        val finalPrice = amount * priceUnit
        val evento = ProductTradeEvent(player)
        evento.product = product
        evento.amount = amount
        evento.newStock = product.stock - amount
        evento.balance = currency!!.get(fake)
        evento.type = TradeType.SELABLE
        evento.priceTotal = finalPrice
        evento.shop = this@Shop
        //
        Mine.callEvent(evento)
        if (evento.isCancelled) {
            return
        }
        if (evento.amount > PLAYER_INVENTORY_LIMIT) {
            evento.amount = PLAYER_INVENTORY_LIMIT.toDouble()
        }
        // quando uma compra é feita seu stock é alterado
        product.stock = evento.newStock
        Mine.remove(player.inventory, product.product, evento.amount.toInt())
        player.sendMessage(
            messageSoldItem.replace("%amount", "" + amount)
                .replace(
                    "%product", product.display
                )
        )
        currency?.add(fake, finalPrice)
    }


    fun getProduct(icon: ItemStack, player: Player): Product? {
        val button = getButton(icon, player)
        if (button != null) {
            if (button is Product) {
                return button
            }
        }
        return null
    }

    fun getProductFrom(item: ItemStack): Product? {
        for (button in buttons)
            if (button is Product) {
                if (true == button.product?.isSimilar(item))
                    return button
            }

        return null
    }

    override fun register(plugin: IPluginInstance) {
        super.register(plugin)
        menuUpgrades?.register(plugin)
        menuConfirmation?.register(plugin)
        if (menuConfirmation != null) {
            val menuConfirmed = menuConfirmation!!
            val confirmationButton = menuConfirmed.getButton("confirmar")!!
            val cancelButton = menuConfirmed.getButton("cancelar")!!
            val productButton = menuConfirmed.getButton("product")!!
            cancelButton.click = ClickEffect { event -> open(event.player) }
            confirmationButton.click = ClickEffect { event ->
                val player = event.player
                val produto = selectedProduct[player]!!
                val type = trading[player]!!

                if (type == TradeType.BUYABLE) {
                    buy(player, produto, 1.0)
                } else if (type == TradeType.SELABLE) {
                    sell(player, produto, 1.0)
                }
                open(event.player)
            }
            if (menuConfirmation!!.openHandler == null)
                menuConfirmation!!.openHandler = { inventory, player ->
                    val product = selectedProduct[player]!!
                    inventory.setItem(productButton.index, product.icon)
                }
        }
        if (menuUpgrades != null) {
            val upgrading = menuUpgrades!!
            upgrading.superiorMenu = this
            if (upgrading.openHandler == null) {
                upgrading.openHandler = { inventory, player ->
                    val product = selectedProduct[player]!!
                    var slot = Extra.getIndex(3, 4)
                    val productButton = menuUpgrades?.getButton("product")!!
                    inventory.setItem(productButton.index, product.icon)
                    for (upgrade in product.upgrades) {
                        val icon = ItemBuilder(Material.STAINED_GLASS_PANE)
                            .data(14)
                            .name("§6Upgrade §e" + upgrade.displayName)
                        if (upgrade.hasBought(player)) {
                            icon.data(5)
                        }
                        inventory.setItem(slot, icon)
                        slot++
                    }
                }
            }
        }
        val upgradeButton = menuUpgrades?.getButton("upgrade") ?: return
        upgradeButton.click = ClickEffect { event ->
            val player = event.player
            val fake = FakePlayer(player)
            val product = selectedProduct[player]!!
            val nextUpgrade = product.getNextUpgrade(player)
            val currency = currency!!
            if (nextUpgrade == null) {
                player.sendMessage(messageUpgradesAlreadyBought)
                return@ClickEffect
            }
            if (!currency.contains(fake, nextUpgrade.price)) {
                player.sendMessage(messageWithoutBalance)
                return@ClickEffect
            }
            currency.remove(fake, nextUpgrade.price)
            VaultAPI.getPermission().playerAdd(player, nextUpgrade.permission)
            player.sendMessage(
                messageUpgradeBought
                    .replace(
                        "%product_name",
                        nextUpgrade.displayName
                    )
                    .replace(
                        "%product",
                        nextUpgrade.name
                    )
                    .replace(
                        "%level",
                        "" + nextUpgrade.level
                    )
            )

        }
    }

    init {

        this.effect = ClickEffect { event ->
            val player = event.player
            if (event.currentItem == null) return@ClickEffect
            val product = getProduct(event.currentItem, player)
            if (product == null) {
                debug("Produto pelo Icone nulo")
                return@ClickEffect
            }

            if (menuUpgrades != null && isPermissionShop && product.hasBought(player)) {
                selectedProduct[player] = product
                menuUpgrades?.open(player)
                return@ClickEffect
            } else if (menuConfirmation != null) {
                if (isPermissionShop && product.hasBought(player)) {
                    return@ClickEffect
                }
                trading[player] = product.tradeType
                selectedProduct[player] = product
                player.closeInventory()
                menuConfirmation!!.open(player)
                return@ClickEffect
            }

            if (((event.click == ClickType.RIGHT || event.click == ClickType.SHIFT_RIGHT) &&
                        (product.tradeType === TradeType.BOTH)) ||
                (product.tradeType === TradeType.BUYABLE)
            ) {
                if (isAmountPerChat && event.click.name.contains("SHIFT")) {
                    selectingAmount[player] = product
                    trading[player] = TradeType.BUYABLE
                    player.closeInventory()
                    player.sendMessage(
                        messageChoiceAmount.replace(
                            "%product",
                            product.display
                        ).replace("%trader", "comprar")
                    )
                    return@ClickEffect
                }

                var amount = product.amount
                if (event.click.name.contains("SHIFT")) {
                    if (amount < 64) {
                        amount = 64
                    }else {
                        amount = 64 * 16
                    }
                    if (product.isLimited && amount > product.stock) {
                        amount = product.stock.toInt()
                    }
                }
                buy(player, product, amount.toDouble())
            }
            if (((event.click == ClickType.LEFT || event.click == ClickType.SHIFT_LEFT)
                        && (product.tradeType === TradeType.BOTH)
                        || product.tradeType === TradeType.SELABLE)
            ) {
                if (product.product == null) return@ClickEffect
                if (isAmountPerChat && event.click.name.contains("SHIFT")) {
                    selectingAmount[player] = product
                    trading[player] = TradeType.SELABLE
                    player.closeInventory()
                    player.sendMessage(
                        messageChoiceAmount
                            .replace("%product", product.display)
                            .replace("%trader", "vender")
                    )
                    return@ClickEffect
                }
                var amount = Mine.getTotalAmount(player.inventory, product.product)
                if (event.click == ClickType.LEFT || event.click == ClickType.LEFT) {
                    if (amount > 64) {
                        amount = 64
                    }
                }
                sell(player, product, amount.toDouble())
            }

        }
    }


}